### PR TITLE
feat: add diagnostic filtering

### LINKED ISSUE #55: [FEATURE REQUEST] Add an option to disable typescript-language-server hints
### Issues

- [X] I have checked [existing issues](https://github.com/jose-elias-alvarez/nvim-lsp-ts-utils/issues) and there are no existing ones with the same request.

### Feature description

typescript-language-server in combination with a linter such as ESLint results in duplicated diagnostics, such as the diagnostic about unused variables. Both VSCode and coc offer such an option, but it is not yet built in typescript-language-server (see: https://github.com/neovim/nvim-lspconfig/issues/997).

### Help

Yes, but I don't know how to start. I would need guidance

### Implementation help

I have made a crude implementation in my config:

```lua
lspconfig.typescript.setup {
  on_attach = function(client, bufnr)
    on_attach(client, bufnr)

    client.handlers['textDocument/publishDiagnostics'] =
      function(err, method, result, client_id, buf_nr, config)
        local filtered_diagnostics = {}
        for _, diagnostic in ipairs(result.diagnostics) do
          if diagnostic.source ~= 'typescript' or diagnostic.severity ~= 4 then
            table.insert(filtered_diagnostics, diagnostic)
          end
        end
        result.diagnostics = filtered_diagnostics

        return vim.lsp.diagnostic.on_publish_diagnostics(
          err,
          method,
          result,
          client_id,
          buf_nr,
          config
        )
      end
  end,
}
```

I'm not sure where I would need to put it. `client.lua` seems to be the most fitting place. Maybe checking for an appropriate option inside `require('nvim-lsp-ts-utils.client').setup` and then setting the wrapper would be a good idea?

### PR DESCRIPTION
**Pull request by nawordar**
Closes #55.

### CODE REVIEW

**Comment by jose-elias-alvarez** [comment]
Thanks for putting this together! I quickly scanned it and it looks good (I especially appreciate the test coverage), but I'll have time to fully review it and test it out locally tomorrow. A couple of quick impressions:

1. I think it would be nice to allow users to set the severities they want to ignore as strings (e.g. `"hint"` instead of `4`), since it's not immediately obvious what the numbers correspond to without referring to the LSP documentation. 
2. Could we mention the option in the readme?

**Comment by nawordar** [comment]
What about having an enum-like table of severities? If so, what are your naming conventions here?

```lua
local ts_utils = require("nvim-lsp-ts-utils")
print(ts_utils.severity.hint) -- 4
```

File: lua/nvim-lsp-ts-utils/client.lua [hunk 1]
```diff
@@ -36,6 +39,36 @@ local edit_handler = function(_, _, workspace_edit)
     return { applied = status, failureReason = result }
 end
 
+local diagnostics_handler = function(err, method, result, client_id, buf_nr, config)
+    local forbidden_severities = o.get().filter_out_diagnostics_by_severity
```
Comment (jose-elias-alvarez) [inline lua/nvim-lsp-ts-utils/client.lua:43]: Could we rename these variables to match the names of the options? I had a moment of confusion testing this out.

File: lua/nvim-lsp-ts-utils/client.lua [hunk 2]
```diff
@@ -36,6 +39,36 @@ local edit_handler = function(_, _, workspace_edit)
...
+                return false
+            end
+
+            return true
+        end, result.diagnostics)
+
+        result.diagnostics = filtered_diagnostics
+    end
+
+    return vim.lsp.diagnostic.on_publish_diagnostics(err, method, result, client_id, buf_nr, config)
```
Comment (jose-elias-alvarez) [inline lua/nvim-lsp-ts-utils/client.lua:69]: We should use `lsp.handlers["textDocument/publishDiagnostics"]` to respect options users have set with `vim.lsp.with` (e.g. disabling virtual text).

**Review by jose-elias-alvarez — CHANGES_REQUESTED** [review]

**Comment by jose-elias-alvarez** [comment]
I'm not opposed to a table of severities, but what's the advantage of having a dedicated variable vs. allowing strings and converting them with a simple map like this?

```lua
local severities = {
    ["error"] = 1,
    ["warning"] = 2,
    -- etc
}
```

For the tests: I think the first test failure is because `diagnostics_result` is getting mutated in the first test. This diff solves it on my end:

```diff
diff --git a/test/spec/client_spec.lua b/test/spec/client_spec.lua
index 9ebb6a2..9b1ca3e 100644
--- a/test/spec/client_spec.lua
+++ b/test/spec/client_spec.lua
@@ -97,14 +97,18 @@ describe("client", function()
         end)

         describe("diagnostics_handler", function()
-            local diagnostics_result = {
-                diagnostics = {
-                    { source = "eslint", severity = 4, code = 80001 },
-                    { source = "typescript", severity = 1, code = 80001 },
-                    { source = "typescript", severity = 3, code = 80001 },
-                    { source = "typescript", severity = 4, code = 80000 },
-                },
-            }
+            local diagnostics_result
+            before_each(function()
+                diagnostics_result = {
+                    diagnostics = {
+                        { source = "eslint", severity = 4, code = 80001 },
+                        { source = "typescript", severity = 1, code = 80001 },
+                        { source = "typescript", severity = 3, code = 80001 },
+                        { source = "typescript", severity = 4, code = 80000 },
+                    },
+                }
+            end)

             it("should filter out hints", function()
                 o.get.returns({
```

I wouldn't worry too much about the import all test, since it's flaky, but you're correct that filtering out errors does affect `:TSLspImportAll`. I don't know if users will really want to filter out errors, so if you think it's okay, I think we can just add a warning in the documentation about the interaction and deal with it later if it comes up.

**Comment by nawordar** [comment]
> I'm not opposed to a table of severities, but what's the advantage of having a dedicated variable vs. allowing strings and converting them with a simple map like this?
> 
> ```lua
> local severities = {
>     ["error"] = 1,
>     ["warning"] = 2,
>     -- etc
> }
> ```

This is a nice idea. I think I will put `severities` in `utils` if it will ever be needed in other files.

**Comment by jose-elias-alvarez** [comment]
LGTM, thanks for working on this! 

I have to think about how to make testing in general more reliable, since Plenary / Luassert weren't really built to handle cases like this, but as long as it actually works it should be fine for now.

**Pull request merged**